File: backend/app/services/skills_service.py
import json
import os
from typing import List, Set
from difflib import SequenceMatcher

class SkillsDatabase:
    def __init__(self):
        self.skills = self._load_skills()
        self.skills_lower = {s.lower(): s for s in self.skills}
    
    def _load_skills(self) -> List[str]:
        """Load skills from skills.json"""
        skills_path = os.path.join(
            os.path.dirname(__file__),
            "skills.json"
        )
        
        try:
            with open(skills_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                return data if isinstance(data, list) else data.get('skills', [])
        except Exception as e:
            print(f"Error loading skills: {e}")
            return []
    
    def get_all_skills(self) -> List[str]:
        """Return all available skills"""
        return sorted(self.skills)
    
    def search_skills(self, query: str, limit: int = 10) -> List[str]:
        """Search skills by name"""
        query_lower = query.lower()
        results = [
            skill for skill in self.skills
            if query_lower in skill.lower()
        ]
        return sorted(results)[:limit]
    
    def normalize_skill(self, skill: str) -> str:
        """Normalize skill name to match database"""
        if not skill:
            return ""
        
        skill_lower = skill.lower().strip()
        
        # Exact match
        if skill_lower in self.skills_lower:
            return self.skills_lower[skill_lower]
        
        # Fuzzy match (similarity > 80%)
        best_match = None
        best_ratio = 0.0
        
        for db_skill_lower, db_skill in self.skills_lower.items():
            ratio = SequenceMatcher(None, skill_lower, db_skill_lower).ratio()
            if ratio > best_ratio:
                best_ratio = ratio
                best_match = db_skill
        
        # Return fuzzy match if similarity > 0.8 (80%)
        if best_ratio > 0.8:
            return best_match
        
        # Return original if no good match
        return skill
    
    def standardize_skills(self, skills: List[str]) -> List[str]:
        """Normalize and deduplicate skills"""
        if not skills:
            return []
        
        standardized = set()
        for skill in skills:
            normalized = self.normalize_skill(skill)
            if normalized:
                standardized.add(normalized)
        
        return sorted(list(standardized))

# Create singleton instance
skills_db = SkillsDatabase()

File: frontend/src/services/skillsService.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

const api = axios.create({
  baseURL: API_BASE_URL,
})

export const skillsService = {
  // Get all available skills
  getAvailableSkills: async () => {
    try {
      const response = await api.get('/api/v1/skills/available')
      return {
        success: true,
        data: response.data.skills
      }
    } catch (error) {
      return {
        success: false,
        error: 'Failed to load skills'
      }
    }
  },

  // Search skills
  searchSkills: async (query: string) => {
    try {
      const response = await api.get('/api/v1/skills/search', {
        params: { query }
      })
      return {
        success: true,
        data: response.data.results
      }
    } catch (error) {
      return {
        success: false,
        error: 'Failed to search skills'
      }
    }
  },

  // Normalize skills
  normalizeSkills: async (skills: string[]) => {
    try {
      const response = await api.post('/api/v1/skills/normalize', skills)
      return {
        success: true,
        data: response.data.standardized
      }
    } catch (error) {
      return {
        success: false,
        error: 'Failed to normalize skills'
      }
    }
  },

  // Get categories
  getCategories: async () => {
    try {
      const response = await api.get('/api/v1/skills/categories')
      return {
        success: true,
        data: response.data
      }
    } catch (error) {
      return {
        success: false,
        error: 'Failed to load categories'
      }
    }
  }
}
